process/block/shardblock.go

```diff
@@ -1588,3 +1595,59 @@ func (sp *shardProcessor) IsInterfaceNil() bool {
 	}
 	return false
 }
+
+func (sp *shardProcessor) addProcessedMiniBlock(metaBlockHash []byte, miniBlockHash []byte) {
+	sp.mutProcessedMiniBlocks.Lock()
+	sp.processedMiniBlocks[string(metaBlockHash)] = append(sp.processedMiniBlocks[string(metaBlockHash)], miniBlockHash)
+	sp.mutProcessedMiniBlocks.Unlock()
+}
+
+func (sp *shardProcessor) removeProcessedMiniBlock(metaBlockHash []byte, miniBlockHash []byte) {
+	var preservedMiniBlockHashes [][]byte
+
+	sp.mutProcessedMiniBlocks.RLock()
+	miniBlockHashes, ok := sp.processedMiniBlocks[string(metaBlockHash)]
+	sp.mutProcessedMiniBlocks.RUnlock()
+
+	if !ok {
+		return
+	}
+
+	for _, hash := range miniBlockHashes {
+		if bytes.Equal(hash, miniBlockHash) {
+			continue
+		}
+
+		preservedMiniBlockHashes = append(preservedMiniBlockHashes, hash)
```

Comment:
I think you can still have some race here.

Reply:
Resolved them